File: test/policy.test.js
import { test } from 'node:test';
import assert from 'node:assert/strict';
import { decide_storage, freshness, may_serve, request_is_cacheable } from '../src/runtime/policy.js';
import { is_authenticated, lookup_keys, normalise_query } from '../src/runtime/key.js';

const config = {
	defaultTtl: 0,
	defaultSwr: 0,
	maxTtl: 86400,
	varyHeaders: []
};

/** A context with nothing set — what a route that never calls platform.cache gets. */
function ctx(overrides = {}) {
	return { tags: new Set(), noStore: false, public: false, segment: undefined, ...overrides };
}

function res(headers = {}, status = 200) {
	return new Response('<html>hi</html>', { status, headers });
}

// ---------------------------------------------------------------------------
// The rules that stop one user seeing another user's page.
// ---------------------------------------------------------------------------

test('an authenticated request is never stored by default', () => {
	const decision = decide_storage({
		response: res({ 'cache-control': 's-maxage=60' }),
		authenticated: true,
		ctx: ctx(),
		config
	});
	assert.equal(decision, null);
});

test('public() is refused for an authenticated request', () => {
	// A route claiming to be identical for everyone cannot be verified once it
	// has rendered with a logged-in user's data, so the claim is declined rather
	// than trusted.
	const decision = decide_storage({
		response: res({ 'cache-control': 's-maxage=60' }),
		authenticated: true,
		ctx: ctx({ public: true }),
		config
	});
	assert.equal(decision, null);
});

test('public() is honoured for an anonymous request', () => {
	const decision = decide_storage({
		response: res({ 'cache-control': 's-maxage=60' }),
		authenticated: false,
		ctx: ctx({ public: true }),
		config
	});
	assert.equal(decision.segment, 'pub');
	assert.equal(decision.ttl, 60);
});

test('an anonymous request lands in the anon segment, not pub', () => {
	const decision = decide_storage({
		response: res({ 'cache-control': 's-maxage=60' }),
		authenticated: false,
		ctx: ctx(),
		config
	});
	assert.equal(decision.segment, 'anon');
});

test('shared() requires an authenticated request', () => {
	assert.equal(
		decide_storage({
			response: res({ 'cache-control': 's-maxage=60' }),
			authenticated: false,
			ctx: ctx({ segment: 'auth' }),
			config
		}),
		null
	);

	assert.equal(
		decide_storage({
			response: res({ 'cache-control': 's-maxage=60' }),
			authenticated: true,
			ctx: ctx({ segment: 'auth' }),
			config
		}).segment,
		'auth'
	);
});

test('a response setting a cookie is never stored', () => {
	// This is what catches the token-refresh path in hooks.server.ts without the
	// route having to know anything about caching.
	const decision = decide_storage({
		response: res({ 'cache-control': 's-maxage=60', 'set-cookie': 'auth_token=new' }),
		authenticated: false,
		ctx: ctx({ public: true }),
		config
	});
	assert.equal(decision, null);
});

test('cache-control private and no-store are refused', () => {
	for (const value of ['private, s-maxage=60', 'no-store, s-maxage=60']) {
		assert.equal(
			decide_storage({
				response: res({ 'cache-control': value }),
				authenticated: false,
				ctx: ctx({ public: true }),
				config
			}),
			null,
			value
		);
	}
});

test('Vary on cookie, authorization or * is refused', () => {
	for (const value of ['Cookie', 'cookie, accept-encoding', 'Authorization', '*']) {
		assert.equal(
			decide_storage({
				response: res({ 'cache-control': 's-maxage=60', vary: value }),
				authenticated: false,
				ctx: ctx({ public: true }),
				config
			}),
			null,
			value
		);
	}
});

test('Vary on an unkeyed header is refused, on a keyed one allowed', () => {
	const response = res({ 'cache-control': 's-maxage=60', vary: 'Accept-Language' });

	assert.equal(
		decide_storage({ response, authenticated: false, ctx: ctx({ public: true }), config }),
		null,
		'not in varyHeaders'
	);

	assert.equal(
		decide_storage({
			response: res({ 'cache-control': 's-maxage=60', vary: 'Accept-Language' }),
			authenticated: false,
			ctx: ctx({ public: true }),
			config: { ...config, varyHeaders: ['accept-language'] }
		}).segment,
		'pub'
	);
});

test('accept-encoding may be varied on without being keyed', () => {
	// Bodies are stored decoded, so encoding negotiation happens downstream.
	const decision = decide_storage({
		response: res({ 'cache-control': 's-maxage=60', vary: 'Accept-Encoding' }),
		authenticated: false,
		ctx: ctx({ public: true }),
		config
	});
	assert.equal(decision.segment, 'pub');
});

test('non-200 responses are never stored', () => {
	for (const status of [301, 404, 500, 206]) {
		assert.equal(
			decide_storage({
				response: res({ 'cache-control': 's-maxage=60' }, status),
				authenticated: false,
				ctx: ctx({ public: true }),
				config
			}),
			null,
			String(status)
		);
	}
});

test('noStore() overrides everything', () => {
	assert.equal(
		decide_storage({
			response: res({ 'cache-control': 's-maxage=3600' }),
			authenticated: false,
			ctx: ctx({ public: true, ttl: 3600, noStore: true }),
			config
		}),
		null
	);
});

// ---------------------------------------------------------------------------
// Freshness and opt-in
// ---------------------------------------------------------------------------

test('nothing is stored without an explicit TTL when defaultTtl is 0', () => {
	assert.equal(
		decide_storage({ response: res(), authenticated: false, ctx: ctx({ public: true }), config }),
		null
	);
});

test('platform.cache.ttl beats the s-maxage header', () => {
	const decision = decide_storage({
		response: res({ 'cache-control': 's-maxage=10' }),
		authenticated: false,
		ctx: ctx({ public: true, ttl: 300, swr: 600 }),
		config
	});
	assert.equal(decision.ttl, 300);
	assert.equal(decision.swr, 600);
});

test('ttl and swr are clamped to maxTtl', () => {
	const decision = decide_storage({
		response: res(),
		authenticated: false,
		ctx: ctx({ public: true, ttl: 999999, swr: 999999 }),
		config
	});
	assert.equal(decision.ttl, 86400);
	assert.equal(decision.swr, 86400);
});

test('stale-while-revalidate is parsed from the header', () => {
	const decision = decide_storage({
		response: res({ 'cache-control': 's-maxage=60, stale-while-revalidate=600' }),
		authenticated: false,
		ctx: ctx({ public: true }),
		config
	});
	assert.equal(decision.ttl, 60);
	assert.equal(decision.swr, 600);
});

test('freshness moves fresh → stale → expired', () => {
	const entry = { storedAt: Date.now(), ttl: 60, swr: 600 };
	assert.equal(freshness(entry, entry.storedAt), 'fresh');
	assert.equal(freshness(entry, entry.storedAt + 30_000), 'fresh');
	assert.equal(freshness(entry, entry.storedAt + 120_000), 'stale');
	assert.equal(freshness(entry, entry.storedAt + 700_000), 'expired');
});

// ---------------------------------------------------------------------------
// Request eligibility and lookup
// ---------------------------------------------------------------------------

test('only GET and HEAD are cacheable', () => {
	for (const method of ['POST', 'PUT', 'DELETE', 'PATCH']) {
		const request = new Request('https://staging.weeb.vip/', { method });
		assert.equal(request_is_cacheable(request), false, method);
	}
	assert.equal(request_is_cacheable(new Request('https://staging.weeb.vip/')), true);
});

test('an Authorization header is never served from cache', () => {
	const request = new Request('https://staging.weeb.vip/', {
		headers: { authorization: 'Bearer x' }
	});
	assert.equal(request_is_cacheable(request), false);
});

test('may_serve refuses to hand an anon entry to a logged-in request', () => {
	assert.equal(may_serve({ segment: 'anon' }, { authenticated: true }), false);
	assert.equal(may_serve({ segment: 'anon' }, { authenticated: false }), true);
	assert.equal(may_serve({ segment: 'pub' }, { authenticated: true }), true);
	assert.equal(may_serve({ segment: 'auth' }, { authenticated: false }), false);
	assert.equal(may_serve(null, { authenticated: false }), false);
});

test('lookup keys differ by auth state and always try pub first', () => {
	const base = {
		prefix: 'ssr',
		version: '1.0.0',
		host: 'staging.weeb.vip',
		method: 'GET',
		pathname: '/airing',
		query: '',
		varyHash: '0'
	};

	const anon = lookup_keys(base, false);
	const authed = lookup_keys(base, true);

	assert.match(anon[0], /:pub:/);
	assert.match(anon[1], /:anon:/);
	assert.match(authed[1], /:auth:/);
	assert.equal(anon[0], authed[0], 'pub key is shared');
	assert.notEqual(anon[1], authed[1], 'segment keys are distinct');
});

test('auth detection ignores emptied and prefixed cookies', () => {
	const names = ['auth_token', 'access_token'];
	assert.equal(is_authenticated('auth_token=abc', names), true);
	assert.equal(is_authenticated('auth_token=', names), false);
	assert.equal(is_authenticated('auth_token=null', names), false);
	// Word-boundary safety: a different cookie ending in the same name must not
	// count, matching the cookieFromString behaviour in auth-storage.ts.
	assert.equal(is_authenticated('guest_access_token=abc', names), false);
	assert.equal(is_authenticated('', names), false);
});

test('tracking params do not fragment the cache', () => {
	const ignored = ['utm_source', 'gclid'];
	const a = normalise_query(new URLSearchParams('q=naruto&utm_source=x'), ignored);
	const b = normalise_query(new URLSearchParams('utm_source=y&q=naruto&gclid=z'), ignored);
	assert.equal(a, b);
	assert.equal(a, 'q=naruto');
});

test('param order does not change the key', () => {
	assert.equal(
		normalise_query(new URLSearchParams('b=2&a=1'), []),
		normalise_query(new URLSearchParams('a=1&b=2'), [])
	);
});

File: edge/policy_test.go
package main

import (
	"net/http"
	"testing"
)

// Mirrors test/policy.test.js. These are the rules that stop one user seeing
// another user's page, and they now exist in two languages — the adapter
// enforces them for its own writes, this proxy enforces them for every other
// framework. They must not drift.

func testConfig() Config {
	return Config{MaxTTL: 86400, VaryHeaders: nil}
}

func resp(status int, headers map[string]string) *http.Response {
	h := http.Header{}
	for k, v := range headers {
		h.Add(k, v)
	}
	return &http.Response{StatusCode: status, Header: h}
}

func TestAuthenticatedRequestNotStoredByDefault(t *testing.T) {
	d := DecideStorage(resp(200, map[string]string{"Cache-Control": "s-maxage=60"}), true, testConfig())
	if d != nil {
		t.Fatalf("stored an authenticated render in segment %q", d.Segment)
	}
}

func TestPublicRefusedForAuthenticatedRequest(t *testing.T) {
	// An origin claiming a page is public cannot be verified once it has
	// rendered with a logged-in user's data, so the claim is declined.
	d := DecideStorage(
		resp(200, map[string]string{"Cache-Control": "s-maxage=60", HeaderSegment: "pub"}),
		true, testConfig(),
	)
	if d != nil {
		t.Fatalf("honoured pub for an authenticated request: %q", d.Segment)
	}
}

func TestPublicHonouredForAnonymousRequest(t *testing.T) {
	d := DecideStorage(
		resp(200, map[string]string{"Cache-Control": "s-maxage=60", HeaderSegment: "pub"}),
		false, testConfig(),
	)
	if d == nil || d.Segment != SegmentPublic {
		t.Fatalf("expected pub, got %+v", d)
	}
}

func TestAnonymousDefaultsToAnonNotPub(t *testing.T) {
	// Sharing with logged-in users must be opted into explicitly; an origin that
	// says nothing gets the isolated segment.
	d := DecideStorage(resp(200, map[string]string{"Cache-Control": "s-maxage=60"}), false, testConfig())
	if d == nil || d.Segment != SegmentAnon {
		t.Fatalf("expected anon, got %+v", d)
	}
}

func TestSharedSegmentRequiresAuthentication(t *testing.T) {
	if d := DecideStorage(
		resp(200, map[string]string{"Cache-Control": "s-maxage=60", HeaderSegment: "auth"}),
		false, testConfig(),
	); d != nil {
		t.Fatalf("stored auth segment for an anonymous request: %+v", d)
	}

	d := DecideStorage(
		resp(200, map[string]string{"Cache-Control": "s-maxage=60", HeaderSegment: "auth"}),
		true, testConfig(),
	)
	if d == nil || d.Segment != SegmentAuth {
		t.Fatalf("expected auth, got %+v", d)
	}
}

func TestSetCookieNeverStored(t *testing.T) {
	d := DecideStorage(
		resp(200, map[string]string{
			"Cache-Control": "s-maxage=60",
			HeaderSegment:   "pub",
			"Set-Cookie":    "auth_token=fresh; Path=/",
		}),
		false, testConfig(),
	)
	if d != nil {
		t.Fatal("stored a response that sets a cookie")
	}
}

func TestPrivateAndNoStoreRefused(t *testing.T) {
	for _, cc := range []string{"private, s-maxage=60", "no-store, s-maxage=60"} {
		if d := DecideStorage(
			resp(200, map[string]string{"Cache-Control": cc, HeaderSegment: "pub"}),
			false, testConfig(),
		); d != nil {
			t.Errorf("stored %q", cc)
		}
	}
}

func TestVaryRefusals(t *testing.T) {
	for _, vary := range []string{"Cookie", "cookie, accept-encoding", "Authorization", "*", "Accept-Language"} {
		if d := DecideStorage(
			resp(200, map[string]string{"Cache-Control": "s-maxage=60", HeaderSegment: "pub", "Vary": vary}),
			false, testConfig(),
		); d != nil {
			t.Errorf("stored despite Vary: %s", vary)
		}
	}
}

func TestVaryOnKeyedHeaderAllowed(t *testing.T) {
	cfg := testConfig()
	cfg.VaryHeaders = []string{"accept-language"}

	d := DecideStorage(
		resp(200, map[string]string{"Cache-Control": "s-maxage=60", HeaderSegment: "pub", "Vary": "Accept-Language"}),
		false, cfg,
	)
	if d == nil {
		t.Fatal("refused a Vary on a keyed header")
	}
}

func TestVaryOnAcceptEncodingAllowed(t *testing.T) {
	// Bodies are stored decoded; encoding is negotiated downstream.
	d := DecideStorage(
		resp(200, map[string]string{"Cache-Control": "s-maxage=60", HeaderSegment: "pub", "Vary": "Accept-Encoding"}),
		false, testConfig(),
	)
	if d == nil {
		t.Fatal("refused Vary: Accept-Encoding")
	}
}

func TestNonOKNeverStored(t *testing.T) {
	for _, status := range []int{301, 404, 500, 206, 304} {
		if d := DecideStorage(
			resp(status, map[string]string{"Cache-Control": "s-maxage=60", HeaderSegment: "pub"}),
			false, testConfig(),
		); d != nil {
			t.Errorf("stored status %d", status)
		}
	}
}

func TestAdapterManagedResponsesAreSkipped(t *testing.T) {
	// The adapter already wrote this entry with tags and a segment the proxy
	// cannot infer. Storing again would produce a competing, dumber copy.
	d := DecideStorage(
		resp(200, map[string]string{
			"Cache-Control": "s-maxage=60",
			HeaderOrigin:    "adapter-knative",
		}),
		false, testConfig(),
	)
	if d != nil {
		t.Fatal("double-stored an adapter-managed response")
	}
}

func TestFreshnessMustBeExplicit(t *testing.T) {
	// No s-maxage and no X-Cache-TTL: an origin that says nothing gets nothing
	// cached, matching the adapter's defaultTtl of 0.
	if d := DecideStorage(resp(200, map[string]string{HeaderSegment: "pub"}), false, testConfig()); d != nil {
		t.Fatal("cached a response with no freshness directive")
	}
}

func TestExplicitHeadersBeatCacheControl(t *testing.T) {
	d := DecideStorage(
		resp(200, map[string]string{
			"Cache-Control": "s-maxage=10, stale-while-revalidate=20",
			HeaderTTL:       "300",
			HeaderSWR:       "600",
			HeaderSegment:   "pub",
		}),
		false, testConfig(),
	)
	if d == nil || d.TTL != 300 || d.SWR != 600 {
		t.Fatalf("expected 300/600, got %+v", d)
	}
}

func TestTTLClampedToMax(t *testing.T) {
	d := DecideStorage(
		resp(200, map[string]string{HeaderTTL: "999999", HeaderSWR: "999999", HeaderSegment: "pub"}),
		false, testConfig(),
	)
	if d == nil || d.TTL != 86400 || d.SWR != 86400 {
		t.Fatalf("expected clamping to 86400, got %+v", d)
	}
}

func TestStaleWhileRevalidateParsed(t *testing.T) {
	d := DecideStorage(
		resp(200, map[string]string{"Cache-Control": "s-maxage=60, stale-while-revalidate=600", HeaderSegment: "pub"}),
		false, testConfig(),
	)
	if d == nil || d.TTL != 60 || d.SWR != 600 {
		t.Fatalf("expected 60/600, got %+v", d)
	}
}

func TestTagsParsed(t *testing.T) {
	h := http.Header{}
	h.Set("Cache-Control", "s-maxage=60")
	h.Set(HeaderSegment, "pub")
	h.Add(HeaderTag, "show:123, airing")
	h.Add(HeaderTag, "home")

	d := DecideStorage(&http.Response{StatusCode: 200, Header: h}, false, testConfig())
	if d == nil {
		t.Fatal("refused")
	}
	want := []string{"show:123", "airing", "home"}
	if len(d.Tags) != len(want) {
		t.Fatalf("got %v want %v", d.Tags, want)
	}
	for i := range want {
		if d.Tags[i] != want[i] {
			t.Fatalf("got %v want %v", d.Tags, want)
		}
	}
}

func TestControlHeadersNotStored(t *testing.T) {
	h := http.Header{}
	h.Set("Content-Type", "text/html")
	h.Set("Set-Cookie", "a=b")
	h.Set("Content-Encoding", "gzip")
	h.Set("Content-Length", "123")
	h.Set(HeaderTag, "x")
	h.Set(HeaderSegment, "pub")
	h.Set(HeaderOrigin, "adapter")

	stored := storableHeaders(h)

	for _, banned := range []string{"Set-Cookie", "Content-Encoding", "Content-Length", HeaderTag, HeaderSegment, HeaderOrigin} {
		if _, present := stored[banned]; present {
			t.Errorf("%s must not be stored", banned)
		}
	}
	if stored["Content-Type"] != "text/html" {
		t.Error("content-type should survive")
	}
}

func TestRequestCacheable(t *testing.T) {
	cases := []struct {
		method  string
		headers map[string]string
		want    bool
	}{
		{"GET", nil, true},
		{"HEAD", nil, true},
		{"POST", nil, false},
		{"PUT", nil, false},
		{"DELETE", nil, false},
		{"GET", map[string]string{"Authorization": "Bearer x"}, false},
		{"GET", map[string]string{"Cache-Control": "no-cache"}, false},
		{"GET", map[string]string{"Cache-Control": "no-store"}, false},
		{"GET", map[string]string{"Upgrade": "websocket"}, false},
	}

	for _, c := range cases {
		r := &http.Request{Method: c.method, Header: http.Header{}}
		for k, v := range c.headers {
			r.Header.Set(k, v)
		}
		if got := RequestCacheable(r); got != c.want {
			t.Errorf("%s %v: got %v want %v", c.method, c.headers, got, c.want)
		}
	}
}
